### DISCUSSION TITLE
Fast data retrieval

### DISCUSSION THREAD

**Discussion by iliavits** [jankae/LibreVNA#259] [General]
I would like to get S11, S12, S21 and S22 as fast as possible from LibreVNA. I am using version 1.5.1 of the GUI and it seems that I can get a sweep of about 500 frequencies for all traces at about 0.23 secs including graphics. When I try to use my own data retrieval program in Python I am getting it at 1.28 secs. I have tried to use https://github.com/jankae/LibreVNA/blob/master/Documentation/UserManual/SCPI_Examples/retrieve_trace_data.py 
but it generates an error:  Expected numeric response from *ESR? but got ''  when  vna.cmd(":DEV:CONN") is executed.
Do you have a working example for fast retrieving S11, S12, S21 and S22 in Python?
Thank you.

**Comment by jankae** [comment]
See #257 for the error message. TLDR: you are using the 1.5.1 release of the GUI (which does not include the *ESR? command) in combination with the python scripting examples from master. This is not a working combination.
You need to either switch to the latest development version for the GUI or stick to the python example scripts from the v1.5.1 tag.

You can get data as fast as it is displayed in the GUI but only for continuous sweeps (single sweeps always add delay due to the setup of the sweep in the FPGA). And the problem when reading data from continuous sweeps is that the SCPI interface is not synchronized to the sweep. Whenever you read the data you just get the latest data of all traces. Compared to your last data read, only a portion of that trace will have changed. Or, if you are reading slower than a whole sweep takes, you will miss data from the intermediate sweep between your data reads.

At the moment, there is no better/faster way to extract the data.

**Comment by iliavits** [comment]
Ok, thank you.
